# Comparison of Bioinformatic Approaches for Minimizing Sequence Redundancy in Amplicon Sequencing Data from Oxford Nanopore Technology Platforms Applied to Dengue Virus Research

## Abstract
This study aims to evaluate bioinformatic methods to reduce sequence redundancy in Oxford Nanopore Technology (ONT) amplicon sequencing of microbial communities, with a particular focus on dengue virus. The study addresses the challenge of high error rates by comparing clustering and denoising techniques to improve sequence robustness, which is critical for molecular epidemiology. The objective of this research is to enhance ONT's capability in various fields, with the potential to impact public health strategies in regions affected by dengue due to climate change.

## Introduction
### Dengue Virus
Dengue virus, also known as "break-bone fever," is a viral infection that spreads to humans through Aedes Aegypti or Aedes Albopictus mosquitoes. It is mostly found in tropical and subtropical climates and can be found in regions such as Eastern Mediterranean, American, South-East Asian, Western Pacific and Africa. Dengue epidemics have a significant impact on the health services, families, and economy of affected countries [1]. Though dengue mainly affects tropical regions, it's now rapidly spreading in European countries with warmer weather such as Croatia, France, Israel, Italy, Portugal, and Spain [2]. This mosquito-borne disease can cause fever, headaches, fatigue, and can be fatal, with up to 40,000 deaths reported each year [3]. Most of the cases or small outbreaks occur due to travelers who are infected abroad and bring the virus back. However, this year, warm weather conditions and an increase in the number of imported cases have led to a surge in local infections carried by tiger mosquitoes, which inhabit southern Europe [4].

![Dengue virus disease case February 2023 - January 2024](images/dengue_virus_map.png)
<p align="center">
  Dengue virus disease case February 2023 - January 2024
</p>

There are four viruses that are related to dengue, which are called serotypes 1-4. Each of these viruses is grouped into genotypes and can cause a variety of clinical symptoms. The severity of the illness can depend on the order in which the person was infected by these viruses. Dengue is classified into three categories: dengue with warning signs, dengue without warning signs, and severe dengue. These viruses belong to the Flavivirus genus of the Flaviviridae family, and they are enveloped, spherical particles with a diameter of approximately 500 Ã…20. Each serotype has a genome of approximately 11 kb of positive-sense, single-stranded RNA that encodes ten proteins. The genome contains three structural proteins, namely, the membrane (M) protein, envelope (E) protein, and capsid (C) protein. It also encodes non-structural (NS) proteins, which are NS1, NS2A, NS2B, NS3, NS4A, NS4B, and NS5. [1] [6]

**Table 1: Dengue Virus Serotypes and Their Genetic Information**
| Virus name       | Accession number | Virus abbreviation |
|------------------|------------------|--------------------|
| Dengue virus 1   | U88536           | DENV-1             |
| Dengue virus 2   | U87411           | DENV-2             |
| Dengue virus 3   | M93130           | DENV-3             |
| Dengue virus 4   | AF326573         | DENV-4             |


### ONT
Studying the genomics of the dengue virus is challenging due to the difficulty in creating unbiased full-length genomes and sequencing them successfully. The current method involves using PCR and capillary sequencing. However, this method is labor-intensive, costly, and limits studies that require identification of true within-host variants. An alternative approach is paired-end short-read sequencing, which enables detailed characterization of low-frequency within-host single nucleotide polymorphisms. [7]

To overcome these issues, whole genome amplification and sequencing presents a cost-effective and high-throughput alternative. Oxford Nanopore Technologies (ONT) offers a third-generation sequencing technology called the MinION sequencer. It generates long reads (> 2 MB nts) and requires fewer resources than other mainstream NGS platforms. It has been successfully used to sequence RNA viruses during epidemics in remote locations without sophisticated laboratory support. [7]

A recent study utilized ONT technology to sequence the genome of the dengue virus. The results were promising in terms of the cost and speed of the process; however, the accuracy of nanopore sequencing was inferior to Illumina sequencing. In particular, the average quality score of nanopore sequencing was Q 7-9 compared to Q 30-40 for Illumina sequencing. A conservative bootstrap cut-off of 99% was required with the nanopore sequencing data to achieve comparable results to the short-read sequencing. When the coverage per base position surpassed 100, the sequences generated by nanopore exhibited more than 99.5% similarity to paired-end short-read sequencing data at the consensus level. Nevertheless, the amount of input DNA recommended for nanopore sequencing is much higher than that for Illumina sequencing (1000 ng versus 0.2 ng). [7]



## Methods





## Result




## Reference
1.	Nature. 2016. Global spread of dengue virus types: mapping the 70 year history. [Online] Available at: https://www.nature.com/articles/nrdp201655
2.	Nature. 2023. How climate change is making dengue fever worse. [Online] Available at: https://www.nature.com/articles/d41586-023-03407-6
3.	PMC. 2000. Dengue and dengue hemorrhagic fever. [Online] Available at: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC10715994/
4.	WHO. 2023. Disease outbreak news on dengue in EU countries. [Online] Available at: https://www.who.int/emergencies/disease-outbreak-news/item/2023-DON498
5.	ECDC. 2023. Dengue worldwide overview. [Online] Available at: https://www.ecdc.europa.eu/en/dengue-monthly
6. https://ictv.global/report/chapter/flaviviridae/flaviviridae/orthoflavivirus
7. https://www.nature.com/articles/s41598-020-75374-1#Sec2